### PR TITLE
Require unique_ptr to Module::addFunctionType()

### PR DESCRIPTION
**Pull request by chfast**
This fixes the memory leak in WasmBinaryBuilder::readSignatures() caused probably the exception thrown there before the FunctionType object is safe.

This also makes it clear that the Module becomes the owner of the FunctionType objects.

### CODE REVIEW

File: src/asmjs/asm_v_wasm.cpp [hunk 1]
```diff
@@ -99,14 +99,13 @@ FunctionType* ensureFunctionType(std::string sig, Module* wasm) {
     return wasm->getFunctionType(name);
   }
   // add new type
-  auto type = new FunctionType;
+  std::unique_ptr<FunctionType> type{new FunctionType};
```
Comment (kripken) [inline src/asmjs/asm_v_wasm.cpp]: Thanks for the PR!

Overall I agree it makes sense to use unique_ptrs for stuff like this. However, we use `auto* thing = make_unique<Type>()` in other places, so we should be consistent one way or the other with the rest of the code. I don't have a strong opinion either way myself.

File: src/binaryen-c.cpp [hunk 2]
```diff
@@ -190,41 +190,42 @@ void BinaryenModuleDispose(BinaryenModuleRef module) {
...
 
-BinaryenFunctionTypeRef BinaryenAddFunctionType(BinaryenModuleRef module, const char* name, BinaryenType result, BinaryenType* paramTypes, BinaryenIndex numParams) {
+BinaryenFunctionTypeRef BinaryenAddFunctionType(BinaryenModuleRef module, const char* name,
+  BinaryenType result, BinaryenType* paramTypes, BinaryenIndex numParams) {
   auto* wasm = (Module*)module;
-  auto* ret = new FunctionType;
-  if (name) ret->name = name;
-  else ret->name = Name::fromInt(wasm->functionTypes.size());
+  std::unique_ptr<FunctionType> ret{new FunctionType};
+  if (name)
```
Comment (kripken) [inline src/binaryen-c.cpp]: our code conventions require `{` and `}` if the if or if-else bodies are not on the same line.

File: src/wasm.h [hunk 3]
```diff
@@ -786,7 +786,7 @@ class Module {
   Function* getFunctionOrNull(Name name);
   Global* getGlobalOrNull(Name name);
 
-  void addFunctionType(FunctionType* curr);
+  FunctionType* addFunctionType(std::unique_ptr<FunctionType> curr);
```
Comment (kripken) [inline src/wasm.h]: I agree this is better. For consistency we should do the others as well. I wouldn't be opposed to separate PRs, during which we are not consistent in the middle, but only if we're sure we can get to the other PRs quickly.

File: src/asmjs/asm_v_wasm.cpp (hunk 1 above)
Comment (chfast) [reply src/asmjs/asm_v_wasm.cpp]: I've seen this, but `make_unique` is not included in every file I modified. I'm happy to change this accordingly.

File: src/binaryen-c.cpp (hunk 2 above)
Comment (chfast) [reply src/binaryen-c.cpp]: Ok, I will try to change it. But I believe I used clang-format here.

File: src/wasm.h (hunk 3 above)
Comment (chfast) [reply src/wasm.h]: Sure, I can send following PRs changing other methods. Do you mean the methods taking raw pointers only in this class?

Here I wanted to send the set of minimal changes to get the feedback about the proposed changes.

Should I change the `curr` argument name? I don't think it is a good name here.

File: src/asmjs/asm_v_wasm.cpp (hunk 1 above)
Comment (kripken) [reply src/asmjs/asm_v_wasm.cpp]: Might need to add an include for it. I think it's in `src/support/utilities.h`.

File: src/binaryen-c.cpp (hunk 2 above)
Comment (kripken) [reply src/binaryen-c.cpp]: I think we investigated that and found we couldn't do it in clang-format, and only clang-tidy can fix up such things. But I may be thinking of another notation.

File: src/wasm.h (hunk 3 above)
Comment (kripken) [reply src/wasm.h]: It might make sense to split up into multiple PRs, good point.

It might be best to wait for #1678 to land first, before any of this, though - it's another widespread change and the two may have conflicts.

About `curr`, it's short for "current" when there isn't a better name. I don't think it's worth changing as it's very widespread in the codebase.

File: src/asmjs/asm_v_wasm.cpp (hunk 1 above)
Comment (chfast) [reply src/asmjs/asm_v_wasm.cpp]: This is fixed already.

File: src/binaryen-c.cpp (hunk 2 above)
Comment (chfast) [reply src/binaryen-c.cpp]: I reverted the code format change.

File: src/wasm.h (hunk 3 above)
Comment (chfast) [reply src/wasm.h]: I applied requested changes. If you want me to wait for some monster PR touching 315 files to be merged... it does not sound reasonable.

**Comment by chfast** [comment]
Should I still work on this?

File: src/binaryen-c.cpp [hunk 4]
```diff
@@ -190,41 +190,40 @@ void BinaryenModuleDispose(BinaryenModuleRef module) {
 
 // Function types
 
-BinaryenFunctionTypeRef BinaryenAddFunctionType(BinaryenModuleRef module, const char* name, BinaryenType result, BinaryenType* paramTypes, BinaryenIndex numParams) {
+BinaryenFunctionTypeRef BinaryenAddFunctionType(BinaryenModuleRef module, const char* name,
```
Comment (kripken) [inline src/binaryen-c.cpp]: please don't change the indentation (we don't have a strict length limit here)

File: src/binaryen-c.cpp [hunk 5]
```diff
@@ -190,41 +190,40 @@ void BinaryenModuleDispose(BinaryenModuleRef module) {
...
-    std::lock_guard<std::mutex> lock(BinaryenFunctionTypeMutex);
-    wasm->addFunctionType(ret);
-  }
-
   if (tracing) {
     std::cout << "  {\n";
     std::cout << "    BinaryenType paramTypes[] = { ";
     for (BinaryenIndex i = 0; i < numParams; i++) {
-      if (i > 0) std::cout << ", ";
+      if (i > 0)
```
Comment (kripken) [inline src/binaryen-c.cpp]: our style requires `{`, `}` if the if body is on another line

File: src/wasm/wasm.cpp [hunk 6]
```diff
@@ -722,15 +722,17 @@ Global* Module::getGlobalOrNull(Name name) {
...
+FunctionType* Module::addFunctionType(std::unique_ptr<FunctionType> curr) {
   if (!curr->name.is()) {
     Fatal() << "Module::addFunctionType: empty name";
   }
   if (getFunctionTypeOrNull(curr->name)) {
     Fatal() << "Module::addFunctionType: " << curr->name << " already exists";
   }
-  functionTypes.push_back(std::unique_ptr<FunctionType>(curr));
-  functionTypesMap[curr->name] = curr;
+  auto p = curr.get();
```
Comment (kripken) [inline src/wasm/wasm.cpp]: please use `auto* ptr =`

File: src/binaryen-c.cpp (hunk 4 above)
Comment (chfast) [reply src/binaryen-c.cpp]: Yes, you do: https://github.com/WebAssembly/binaryen/blob/master/.clang-format#L4

**Comment by axic** [comment]
@chfast ping

**Comment by chfast** [comment]
> Have you joined the [w3c community group](https://www.w3.org/community/webassembly/participants)? (required for contributors)

I have.



> Looks like there are some compilation errors on CI here.

Fixed.

**Comment by chfast** [comment]
Ping.

File: src/ir/function-type-utils.h [hunk 7]
```diff
@@ -23,9 +23,9 @@ namespace wasm {
 namespace FunctionTypeUtils {
 
 // Fills in function info from a function type
-inline void fillFunction(Function* func, FunctionType* type) {
-  func->params = type->params;
-  func->result = type->result;
+inline void fillFunction(Function& func, const FunctionType& type) {
```
Comment (kripken) [inline src/ir/function-type-utils.h]: Why is the signature changed here? The downside is that we do tend to use pointers in most places, which means a bunch of extra `*`, which I think is less readable.

Comment (chfast) [reply src/ir/function-type-utils.h]: I removed this commit.

**Comment by kripken** [comment]
Great, thanks!
